File: Web/CustomPreview/C#/CustomPreview/RdlxHandlerMiddleware.cs
using GrapeCity.ActiveReports;
using GrapeCity.ActiveReports.Core.Export.Html.Page;
using GrapeCity.ActiveReports.Export.Html.Page;
using GrapeCity.ActiveReports.Extensibility.Rendering;
using GrapeCity.ActiveReports.Rendering.IO;
using Microsoft.Extensions.Caching.Memory;
using Microsoft.Net.Http.Headers;

namespace CustomPreview;

public class RdlxHandlerMiddleware
{
	private const string HandlerExtension = ".rdlx";
	private const string HandlerCacheExtension = ".rdlxWeb";
	
	public RdlxHandlerMiddleware(RequestDelegate next) { }
	
	public async Task InvokeAsync(HttpContext context, IWebHostEnvironment webHostEnvironment)
	{
		var cache = GetCache(context)!;
		if (!context.Request.Path.ToString().EndsWith(HandlerExtension))
		{
			if (!context.Request.Path.ToString().EndsWith(HandlerCacheExtension))
				return;

			// return image
			var keyName = Path.GetFileName(context.Request.Path.ToUriComponent());
			if (cache.TryGetValue(keyName, out var cacheItem))
			{
				var cacheItemByteArray = (byte[]?) cacheItem;
				// For SVG images the automatic detection of Response.ContentType sets the wrong type,
				// so we have to set the content type manually 
				if (keyName.EndsWith(".svg" + HandlerCacheExtension))
					context.Response.ContentType = "image/svg+xml";
				await context.Response.Body.WriteAsync(cacheItemByteArray);
			}
			return;
		}

		context.Response.ContentType = "text/html";
		var streams = new HtmlStreamProvider();
		try
		{
			using (var report = new PageReport(new FileInfo(webHostEnvironment.ContentRootPath + context.Request.Path)))
			{
				var html = new HtmlRenderingExtension();
				var settings = (Settings)html.GetSupportedSettings();
				settings.StyleStream = false;
				settings.OutputTOC = false;
				settings.Mode = RenderMode.Paginated;
				report.Document.Render(html, streams, ((ISettings)settings).GetSettings());
			}

			foreach (var secondaryStreamInfo in streams.GetSecondaryStreams())
			{
				var secondaryStream = (MemoryStream) secondaryStreamInfo.OpenStream();
				// 30 seconds should be enough to load all images
				cache.Set(secondaryStreamInfo.Uri.OriginalString, secondaryStream.ToArray(),
					new MemoryCacheEntryOptions { SlidingExpiration = new TimeSpan(0, 0, 30) });
			}
		}
		catch (ReportException eRunReport)
		{
			// Failure running report, just report the error to the user.
			context.Response.StatusCode = 500;
			await context.Response.WriteAsync(Resources.Resources.Error + eRunReport);
			return;
		}

		var primaryStream = (MemoryStream) streams.GetPrimaryStream().OpenStream();

		context.Response.GetTypedHeaders().CacheControl = new CacheControlHeaderValue { NoCache = true };
		await context.Response.Body.WriteAsync(primaryStream.ToArray());
	}
	
	private IMemoryCache? GetCache(HttpContext context) => context.RequestServices.GetService<IMemoryCache>();
	
	private sealed class HtmlStreamProvider : MemoryStreamProvider
	{
		protected override Uri GetStreamUri(string extension)
		{
			Uri uri = base.GetStreamUri(extension);
			return new Uri(uri.OriginalString + HandlerCacheExtension, UriKind.Relative);
		}
	}
}

File: Web/CustomPreview/VB.NET/CustomPreview/RdlxHandlerMiddleware.vb
Imports System.IO
Imports GrapeCity.ActiveReports
Imports GrapeCity.ActiveReports.Core.Export.Html.Page
Imports GrapeCity.ActiveReports.Export.Html.Page
Imports GrapeCity.ActiveReports.Extensibility.Rendering
Imports GrapeCity.ActiveReports.Extensibility.Rendering.IO
Imports GrapeCity.ActiveReports.Rendering.IO
Imports Microsoft.AspNetCore.Hosting
Imports Microsoft.AspNetCore.Http
Imports Microsoft.Extensions.Caching.Memory
Imports Microsoft.Extensions.DependencyInjection
Imports Microsoft.Net.Http.Headers

Public Class RdlxHandlerMiddleware
    
    Private Const HandlerExtension = ".rdlx"
    Private Const HandlerCacheExtension = ".rdlxWeb"
    
    Public Sub New (rd As RequestDelegate)
    End Sub
    
    Public Async Function InvokeAsync(context as HttpContext, hostingEnv as IWebHostEnvironment) As Task
        Dim cache = GetCache(context)
        If Not context.Request.Path.ToString().EndsWith(HandlerExtension)
            If Not context.Request.Path.ToString().EndsWith(HandlerCacheExtension)
                Return
            End If
            
            ' return image
            Dim keyName = Path.GetFileName(context.Request.Path.ToUriComponent())
            Dim cacheItem As Byte() = Nothing
            If cache.TryGetValue(keyName, cacheItem) Then
                ' For SVG images the automatic detection of Response.ContentType sets the wrong type,
                ' so we have to set the content type manually 
                If keyName.EndsWith(".svg" + HandlerCacheExtension) Then
                    context.Response.ContentType = "image/svg+xml"
                End If
                Await context.Response.Body.WriteAsync(cacheItem)
            End If
            Return
        End If
        
        context.Response.ContentType = "text/html"
        Dim reportException As String = Nothing
        Dim streams = New HtmlStreamProvider()
        Try
            Using report = New PageReport(New FileInfo(hostingEnv.ContentRootPath + context.Request.Path))
                Dim html = New HtmlRenderingExtension()
                Dim settings = DirectCast(html.GetSupportedSettings(), Settings)
                settings.StyleStream = False
                settings.OutputTOC = False
                settings.Mode = RenderMode.Paginated
                report.Document.Render(html, streams, DirectCast(settings, ISettings).GetSettings())
            End Using
            
            For Each secondaryStreamInfo As StreamInfo In streams.GetSecondaryStreams()
                Dim secondaryStream = DirectCast(secondaryStreamInfo.OpenStream(), MemoryStream)
                ' 30 seconds should be enough to load all images
                cache.Set(secondaryStreamInfo.Uri.OriginalString, secondaryStream.ToArray(),
                          New MemoryCacheEntryOptions With { .SlidingExpiration = New TimeSpan(0, 0, 30) })
            Next
            
        Catch eRunReport As ReportException
            ' Failure running report, just report the error to the user.
            reportException = eRunReport.ToString()
        End Try
        ' VB doesn't support Await inside Try Catch block
        If Not String.IsNullOrEmpty(reportException)
	        context.Response.StatusCode = 500
            Await context.Response.WriteAsync(My.Resources.Resources.RunningError + reportException)
            Return
        End If
        
        Dim primaryStream = DirectCast(streams.GetPrimaryStream().OpenStream(), MemoryStream)
	   
	    context.Response.GetTypedHeaders().CacheControl = new CacheControlHeaderValue With { .NoCache = true }
        Await context.Response.Body.WriteAsync(primaryStream.ToArray())
    End Function
    
    Private Function GetCache(context as HttpContext) as IMemoryCache
        Return context.RequestServices.GetService(Of IMemoryCache)
    End Function
    
    Private NotInheritable Class HtmlStreamProvider
        Inherits MemoryStreamProvider
        Protected Overrides Function GetStreamUri(extension As String) As Uri
            Dim uri As Uri = MyBase.GetStreamUri(extension)
            Return New Uri(uri.OriginalString + HandlerCacheExtension, UriKind.Relative)
        End Function
    End Class
End Class
